### DISCUSSION TITLE
how can make a custom pagination api

### DISCUSSION THREAD

**Discussion by Haythamfpco** [laravel/framework#41513] [Q&A] [answered]
hi 
I Have Q: how do I can to customize pagination  like this 

```php
'data' : [
    {
        // Record...
    },
    {
        // Record...
    }
],
'pagination' : [
   "total": 50,
   "count": 50,
   "per_page": 15,
   "current_page": 1,
   "pre_page": 1,
   "total_pages": 15,
],
```

I solve by this solution but I repeat this code in all ResourceCollection so how I can overwrite paginate method or any solution 
```php 
 public function index(): JsonResponse
    {
        $accounts = Account::with([
            'userAdd:id,full_name',
            'userEdit:id,full_name',
        ])
        ->when(request('name') ,fn($q) =>$q->where('name', 'LIKE', '%' . request('name') . '%'))
        ->when(request('phone') ,fn($q) =>$q->where('phone', 'LIKE', '%' . request('phone') . '%'))
        ->orderBy('created_at','desc');
        return $this->getSuccessData(new PaginationAccountResource($accounts->paginate(10)));
    }
```

```php
<?php

namespace App\Http\Resources\Account;

use Illuminate\Http\Resources\Json\ResourceCollection;

class PaginationAccountResource extends ResourceCollection
{

    public function toArray($request) : array
    {

        return [
            'data'       => IndexAccountResource::collection($this->collection),
            'pagination' => [
                'total'       => $this->total(),
                'count'       => $this->count(),
                'perPage'     => $this->perPage(),
                'currentPage' => $this->currentPage(),
                'totalPages'  => $this->lastPage()
            ],
        ];
    }

    public function withResponse($request, $response)
    {
        $jsonResponse = json_decode($response->getContent(), true);
        unset($jsonResponse['links'],$jsonResponse['meta']);
        $response->setContent(json_encode($jsonResponse['data']));
    }

}

```

**Comment by mikeydevelops** [comment]
Its really easy.

First, create a custom paginator that extends `LengthAwarePaginator`:
I made it in `app/Pagination/CustomPaginator.php`:
```php
<?php

namespace App\Pagination;

use Illuminate\Pagination\LengthAwarePaginator;

class CustomPaginator extends LengthAwarePaginator
{
    /**
     * Get the instance as an array.
     *
     * @return array
     */
    public function toArray()
    {
        return [
            'data' => $this->items->toArray(),
            'pagination' => [
                'total'       => $this->total(),
                'count'       => $this->count(),
                'perPage'     => $this->perPage(),
                'currentPage' => $this->currentPage(),
                'totalPages'  => $this->lastPage(),
            ],
        ];
    }
}

```

And in your `App\Providers\AppServiceProvider` (or some other provider, for example custom only for pagination stuff) register the custom paginator like this:
```php

use App\Pagination\CustomPaginator;
use Illuminate\Pagination\LengthAwarePaginator;
use Illuminate\Contracts\Pagination\LengthAwarePaginator as LengthAwarePaginatorContract;

public function register()
{
    $this->app->alias(CustomPaginator::class, LengthAwarePaginator::class); // Eloquent uses the class instead of the contract 🤔
    $this->app->alias(CustomPaginator::class, LengthAwarePaginatorContract::class);
}
```

And now every time a length aware paginator is resolved from the container like Eloquent and Query Builder do, you will receive an instance of your CustomPaginator:
```php
User::paginate();

DB::table('users')->paginate();
```

**Reply by Haythamfpco** [reply]
thank You @mikeydevelops

**Reply by getsolaris** [reply]
thank you @mikeydevelops

**Comment by Haythamfpco** [comment]
@mikeydevelops 

it's work if i used  only paginate  ``DB::table('users')->paginate(); `` 
but I need to  use `IndexAccountResource::collection($accounts->paginate(10))` in this caes the reutrns likes this : 

```
    {
        // Record...
    },
    {
        // Record...
    }

```


IndexAccountResource 


```php
class IndexAccountResource extends JsonResource
{

    public function toArray($request) : array
    {

        return [
            'id'                  => $this->id,
            'name'                => $this->name,
            'manger'              => $this->manger_name,
            'identity'            => $this->identity    ?? '--',
            'gender'              => $this->gender,
            'city'                => $this->city        ?? '--',
            'address'             => $this->address     ?? '--',
            'work'                => $this->field_work  ?? '--',
            'phone'               => $this->phone       ?? '--',
            'accountParent'       => $this->parentAccount->name ?? '--',
            'accountType'         => $this->accountType->name,
            'farisCard'           => implode(' | ',$this->getCards()) ,
            'userAdd'             => $this->userAdd->full_name,
            'userEdit'            => $this->userEdit->full_name ?? '--',
            'createdAt'           => $this->created_at,
            'updatedAt'           => $this->updated_at,
        ];
    }
}
```

**Comment by mikeydevelops** [comment, marked as answer]
I have dug in to the source of Resources and ResourceCollections and come up with a plan.
Together with [my previous response](https://github.com/laravel/framework/discussions/41513#discussioncomment-2374919) and this you will achieve your desired result.

1. First you have to make a base resource, I will call it `BaseResource` for the example, which extends the original `JsonResource`, allowing you to overwrite methods. 
2. Then place the `BaseResource` somewhere I will use `App\Http\Resources\BaseResource`.
3. Go through all your resource classes and change `class ... extends JsonResource` to `class ... extends BaseResource`.
4. Copy the definition of the [`collection`](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Http/Resources/Json/JsonResource.php#L77) method from the `JsonResource` and place it in your `BaseResource`.
5. Rename the `new AnonymousResourceCollection` with `new CustomResourceCollection` in the copied `collection` method.
6. Now make a new class `CustomResourceCollection` which extends the original [`AnonymousResourceCollection`](https://github.com/laravel/framework/blob/9.x/src/Illuminate/Http/Resources/Json/AnonymousResourceCollection.php) which I will also place in `App\Http\Resources\CustomResourceCollection`
7. Add this method to `CustomResourceCollection`:
```php
/**
 * Add the pagination information to the response.
 *
 * @param  \Illuminate\Http\Request  $request
 * @param  array  $paginated
 * @param  array  $default
 * @return array
 */
public function paginationInformation($request, $paginated, $default)
{
    return Arr::only($paginated, 'pagination');
}
```
8. and the final step is to remember to change the `exends` expression for each new resource to extend your custom resource.
9. Alternatively if you are using the command line to create your resources, i.e. `php artisan make:resource` you can publish Laravel's default stubs for the new files and edit the defaults to match with your custom resource and custom resource collection.
    1. Run `php artisan stub:publish`
    2. now you have a new folder called `stubs` in the base of your projects, inside there are many files which Laravel uses to generate your classes.
    3. Find `resource.stub` and `resource-collection.stub` and edit them to match your custom resource.

Thats it, quite a bit of steps, but all is working as you wanted.

**Comment by Haythamfpco** [comment]
@mikeydevelops
thank you so much for your time but it's although  return like this without pagination

```
 {
        // Record...
    },
    {
        // Record...
    }
```

and this is snpit code 
```php 
<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\JsonResource;

class BaseResource extends JsonResource
{

    
    public static function collection($resource)
    {
        return tap(new CustomResourceCollection($resource, static::class), function ($collection) {
            if (property_exists(static::class, 'preserveKeys')) {
                $collection->preserveKeys = (new static([]))->preserveKeys === true;
            }
        });
    }

}


```

```php

<?php

namespace App\Http\Resources;

use Illuminate\Http\Request;
use Illuminate\Http\Resources\Json\AnonymousResourceCollection;
use Illuminate\Support\Arr;

class CustomResourceCollection extends AnonymousResourceCollection
{


    /**
     * Add the pagination information to the response.
     *
     * @param  Request  $request
     * @param  array  $paginated
     * @param  array  $default
     * @return array
     */
    public function paginationInformation($request, $paginated, $default)
    {
        return Arr::only($paginated, 'pagination');
    }

}
```

```php

<?php

namespace App\Http\Resources\Account;

use App\Http\Resources\BaseResource;

class IndexAccountResource extends BaseResource
{

    public function toArray($request) : array
    {

        return [
            'id'                  => $this->id,
            'name'                => $this->name,
            'manger'              => $this->manger_name,
            'identity'            => $this->identity    ?? '--',
            'gender'              => $this->gender,
            'city'                => $this->city        ?? '--',
            'address'             => $this->address     ?? '--',
            'work'                => $this->field_work  ?? '--',
            'phone'               => $this->phone       ?? '--',
            'accountParent'       => $this->parentAccount->name ?? '--',
            'accountType'         => $this->accountType->name,
            'userAdd'             => $this->userAdd->full_name,
            'userEdit'            => $this->userEdit->full_name ?? '--',
            'createdAt'           => $this->created_at,
            'updatedAt'           => $this->updated_at,
        ];
    }
}
```
```php

public function index(): JsonResponse
    {
        $accounts = Account::with([
            'userAdd:id,full_name',
            'userEdit:id,full_name',
        ])
        ->when(request('name') ,fn($q) =>$q->where('name', 'LIKE', '%' . request('name') . '%'))
        ->when(request('phone') ,fn($q) =>$q->where('phone', 'LIKE', '%' . request('phone') . '%'))
        ->orderBy('created_at','desc');
        return $this->getSuccessData( IndexAccountResource::collection($accounts->paginate(10)) );
    
```

**Reply by mikeydevelops** [reply]
Hmm, interesting, can you tell me what version of Laravel are you using, maybe mine is different?
Also when you pass `$accounts->paginate(10)` to `dd` (`dd($accounts->paginate(10))`) what do you get, it should be an instance of `App\PAgination\CustomPaginator`?

My Laravel version is 9.5.1.

Here is my setup and is almost the same:
I am using Users to test.
The UserResource:
![image](https://user-images.githubusercontent.com/10907429/158899199-3fcd6365-c147-4a78-880d-5ed726853872.png)

The BaseResource:
![image](https://user-images.githubusercontent.com/10907429/158899260-080963c8-2561-40b7-a8ac-3aa96d6f73be.png)

The CustomResourceCollection:
![image](https://user-images.githubusercontent.com/10907429/158899331-876df1ae-071c-4a35-8fd0-d7a521fff6d9.png)

The Route Configuration:
![image](https://user-images.githubusercontent.com/10907429/158899422-5be4182f-d46d-428a-a146-ba017ca648ec.png)

And the response from the route:
![image](https://user-images.githubusercontent.com/10907429/158899506-547b7bcc-9f1b-4521-b1af-4ccf1ded93eb.png)

**Comment by Haythamfpco** [comment]
@mikeydevelops 
thank you so match
 the problem on function controller and i solved it   
![image](https://user-images.githubusercontent.com/37311945/158910640-698be417-71cf-4e19-9634-4addf1003255.png)

**Comment by phyothiha** [comment]
@mikeydevelops This solution works like a charm, but when I use the response json method, only data array values will show. Is there any way around it? 

Return a response from a controller with
```php
return response()->json(UserResource::collection(User::paginate(10));
```

The output
```json
{
    "success": true,
    "statusCode": 200,
    "data": [
        {
            "id": 8,
            "name": "John Doe",
            "title": "Nobis esse nisi omni"
        },
        {
            "id": 7,
            "name": "John Doe",
            "title": "Aute ratione hic in"
        }
    ]
}
```

The expected output will be
```json
{
    "success": true,
    "statusCode": 200,
    "data": [
        {
            "id": 8,
            "name": "John Doe",
            "title": "Nobis esse nisi omni"
        },
        {
            "id": 7,
            "name": "John Doe",
            "title": "Aute ratione hic in"
        }
    ],
    "pagination": {
        "total": 8,
        "count": 2,
        "per_page": 2,
        "current_page": 1,
        "last_page": 4
    }
}
```

**Reply by HuyNguyen206** [reply]
In the CustomResourceCollection.php, update to like below


`<?php

namespace App\Http\Resources;

use Illuminate\Http\Resources\Json\AnonymousResourceCollection;
use Illuminate\Support\Arr;

class CustomResourceCollection extends AnonymousResourceCollection
{
    public function paginationInformation($request, $paginated, $default)
    {
        return Arr::only($paginated, ['meta']);
    }
}`

And then try update this one:
`return response()->json(UserResource::collection(User::paginate(10));`
to
`return response()->json(UserResource::collection(User::paginate(10))->response()->getData());`
